### ISSUE TITLE
Add documentation on Angular Libraries best practices

### ISSUE THREAD

**Issue by LAlves91** [angular/angular#42203]
Greetings!

I'm currently working on improvements in the codebase for an Angular Library used in the web apps I work on. When this library was first created, we didn't have a lot of information about libraries, so there are a lot of problems regarding structure and modular approach.

I've been searching for more information on the subject, but I believe the current official documentation on custom libraries is lacking in details, especially concerning best practices for performance (like multiple entry points, configurations and structures). My main source of knowledge on the subject at the moment are some articles on the web and open source repos.

So, the purpose of this issue is to request improvements in the official library documentation, or even some other sources about this subject. I believe this woud be great for the Angular dev community!

Thank you!

**Comment by vugar005** [comment]
Yes it would be very nice. You can also  follow article of @tomastrajan  blog[ https://link.medium.com/2QwTVrOzNgb]()

**Comment by tomastrajan** [comment]
> Yes it would be very nice. You can also follow article of @tomastrajan blog https://link.medium.com/2QwTVrOzNgb

The link seem to point to this issue so here is another one 😉

https://tomastrajan.medium.com/the-best-way-to-architect-your-angular-libraries-87959301d3d3?sk=0b449c5d8dd8f1b84d7957b5af8f4ab4

**Comment by vugar005** [comment]
@tomastrajan  oh sorry lol, I posted this from mobile :)

**Comment by Jadamae77** [comment]
+1

**Comment by aaronmanela** [comment]
+1

**Comment by petebacondarwin** [comment]
We committed a number of changes to the docs about creating libraries in the last three weeks. See https://github.com/angular/angular/commits/master/aio/content/guide/creating-libraries.md

Please can you take a look and let us know if this covers the areas you wanted improved so that we can close this issue as resolved? If it does not, please can you list specific items that need to be addressed so that we can action this request.

**Comment by LAlves91** [comment]
Hi @petebacondarwin !

I think there are some more topics that could be explored (in this doc or maybe in a new "advanced concepts in angular libraries" one):
- How libraries consumption works when a client app is bundled and how to best optimize it (and optimization approaches in cases of plain JS peer dependencies);
- Secondary entry-points (kinda related to the topic above) and how to better make use of them (approaches to separate/put together artifacts in entry-points, folder structure, use of path mapping and barrel import);
- Strategies to avoid circular dependencies (which breaks libraries migrating to Angular 12+);

Those are the topics that come to mind here (which are based on some of my struggles refactoring a gigantic Angular library in my company - more of a core frontend library for multiple related apps, instead of just reusable components). Maybe other participants of this issue have some other suggestions!

Cheers!

**Comment by tomastrajan** [comment]
Hey @LAlves91 

> Secondary entry-points (kinda related to the topic above) 
the secondary entry points are now have official documentation [here](https://angular.io/guide/angular-package-format#entrypoints-and-code-splitting) 👍 

> Strategies to avoid circular dependencies (which breaks libraries migrating to Angular 12+);
(circular deps can be prevented with sub-entries

**Comment by lwensveen** [comment]
I would like to add to this:

I'm trying to migrate a angular app to a multi project workspace where my library already lives. The errors that I now get inside my library are:

`No pipe found with name 'async'.` and `Can't bind to 'ngClass' since it isn't a known property of 'div'.`  which would imply that I'm missing a CommonModule, yet when I check the module imports it's there and the component itself is also in the declarations. As far as I know those are the two things to check when this occurs.

Now this error isn't giving me any useful information and the docs are pretty summary which makes it hard to look for a solution.

**Comment by varthc** [comment]
> the secondary entry points are now have official documentation [here](https://angular.io/guide/angular-package-format#entrypoints-and-code-splitting) 👍

IMHO it'd be nice to add more info/examples about it, instead of just explaining what it is.
